src/content/peerconnection/negotiate-timing/js/main.js

```diff
@@ -0,0 +1,232 @@
+  }
+}
+
+async function onIceCandidate(pc, event) {
+  if (event.candidate) {
+    console.log(`${getName(pc)} emitted ICE candidate for index ${event.candidate.sdpMLineIndex}:\n${event.candidate.candidate}`);
+  } else {
+    console.log(`$getName(pc)} ICE NULL candidate`);
+  }
+  await getOtherPc(pc).addIceCandidate(event.candidate);
+  console.log(`${getName(pc)} addIceCandidate success`);
+}
+
+function onIceStateChange(pc, event) {
+  if (pc) {
+    console.log(`${getName(pc)} ICE state: ${pc.iceConnectionState}`);
+    console.log('ICE state change event, state: ', pc.iceConnectionState);
+  }
+}
+
+function adjustTransceiverCounts(pc, videoCount) {
+  const currentVideoTransceivers = pc.getTransceivers().filter(tr => tr.receiver.track.kind == 'video');
+  const currentVideoCount = currentVideoTransceivers.length;
+  if (currentVideoCount < videoCount) {
+    console.log('Adding ' + (videoCount - currentVideoCount) + ' transceivers');
+    for (let i = currentVideoCount; i < videoCount; ++i) {
+      pc.addTransceiver('video');
+    }
+  } else if (currentVideoCount > videoCount) {
+    console.log('Stopping ' + (currentVideoCount - videoCount) + ' transceivers');
+    for (let i = videoCount; i < currentVideoCount; ++i) {
+      currentVideoTransceivers[i].stop();
+    }
+  } else {
+    console.log(`No adjustment, video count is ${currentVideoCount}, target was ${videoCount}`);
+  }
+}
+
+async function getAudioImpairment(audioTransceiver) {
+  const stats = await audioTransceiver.receiver.getStats();
+  let currentImpairment;
+  stats.forEach(stat => {
+    if (stat.type == 'track') {
+      currentImpairment = stat.concealedSamples;
+    }
+  });
+  console.log('Found impairment value ', currentImpairment);
+  return currentImpairment;
+}
+
+async function baselineAudioImpairment(pc) {
+  audioTransceiver = pc.getTransceivers().filter(tr => tr.receiver.track.kind == 'audio')[0];
+  console.log('Found audio transceiver');
+  audioImpairmentAtStart = await getAudioImpairment(audioTransceiver);
+}
+
+async function measureAudioImpairment(pc) {
+  const startTime = performance.now();
+  const audioImpairmentNow = await getAudioImpairment(audioTransceiver);
+  console.log('Measurement took ' + (performance.now() - startTime) + ' msec');
+  return audioImpairmentNow - audioImpairmentAtStart;
+}
+
+
+async function renegotiate() {
+  adjustTransceiverCounts(pc1, parseInt(videoSectionsField.value));
+  renegotiateButton.disabled = true;
+  await baselineAudioImpairment(pc2);
+  const previousVideoTransceiverCount = pc2.getTransceivers().filter(tr => tr.receiver.track.kind == 'video').length;
+  const startTime = performance.now();
+  const offer = await pc1.createOffer();
+  await pc1.setLocalDescription(offer);
+  await pc2.setRemoteDescription(offer);
+  const answer = await pc2.createAnswer();
+  await pc1.setRemoteDescription(answer);
+  await pc2.setLocalDescription(answer);
+  const elapsedTime = performance.now() - startTime;
```

Comment:
Maybe de-duplicate with the code in call()?

Reply:
May be a good idea. The performance of initial negotiation may be different from later negotiation, but both are interesting numbers.